Add section on IL Stubs. (#2)
AaronRobinsonMSFT/interop-doc: doc/interop-doc.md

    - [Garbage Collection](#gc) &ndash; Pinning and `GCHandle`s.
    - [Referencing Memory](#referencingmemory).
    - [Blittable vs Unmanaged types](#blittablevsunmanaged).

>     - [IL Stubs and Reverse IL Stubs](#ilstubs) &ndash; How stubs are generated and runtime optimizations.

Reviewer: The current phrasing seems awkward to me.

Maybe `Stub generation and runtime optimizations.`?

Reviewer: Also, I don't think the section really talks about 'runtime optimizations' - is that still a todo somewhere?

Author: Remove comment on 'runtime optimizations' - that was premature.

### IL Stubs and Reverse IL Stubs <a name="ilstubs"></a>

The transition between a managed and native environment during functions calls may require help for various reasons. The most obvious cases involve marshalling types that aren't blittable as was [previously illustrated](#transition), but other cases occur when the semantics of the call are altered to improve the interop experience. For example, consider the [`PreserveSig`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.dllimportattribute.preservesig) field on [`DllImportAttribute`][api_dllimportattr]. This field indicates if APIs that return a Windows' [`HRESULT`][wiki_hresult] error code should convert that error code into a .NET [`Exception`][api_exception] or retain the original function signature and return an `int`. This translation is handled in an ad-hoc basis based on the function declaration and is performed by the runtime generated IL Stub.

> Before discussing what these stubs do, lets define what they don't do. The stubs themselves do not perform any GC mode transitions (that is, [switch from Preemptive to Cooperative mode][doc_botr]). That transition logic is performed around the stubs but added by the JIT during JIT compilation. The stubs are also not responsible for handling native calling conventions. Prior to .NET 6, the generated IL Stubs were involved in calling conventions under certain circumstances when calling a COM or C++ member function on Windows. A consequence of having the IL Stubs handle aspects of calling conventions meant that anyone wanting to auto-generate interop code also had to handle these same aspects. As of .NET 5, the IL Stubs also don't do anything that can't be done in C#. Prior to C# function pointers, the IL Stubs used IL instructions that were not representable in C# and therefore were difficult to be source generated.

Reviewer: `lets` > `let's`

Reviewer: > source generated

Is that a normal term? I know we have been saying it (although I envision it as hyphenated), but everything I have seen uses more verbose/descriptive phrasing.

Reviewer: Thoughts on turning the things the don't do (and associated details) into bullets?

I think it might be faster to read/understand, especially with the qualifications like 'prior to .NET 6'.

Let's enumerate what IL Stubs do and what influences them.

- Respond to the fields on the [`DllImportAttribute`][api_dllimportattr]. Each of the available fields influences either the logic in the IL Stub or the intended native target.

> - Respond to interop related attributes that control transitioning or marshalling behavior. For example, see [`PreserveSigAttribute`][api_preservesigattr] and [`UnmanagedFunctionPointerAttribute`][api_unmanagedfunctionpointerattr].

Reviewer: ```suggestion
- Respond to interop-related attributes that control transitioning or marshalling behavior. For example, see [`PreserveSigAttribute`][api_preservesigattr] and [`UnmanagedFunctionPointerAttribute`][api_unmanagedfunctionpointerattr].
```

> - Marshal each argument into the appropriate form for the target. For example, if a .NET `string` is being passed to a native function then by default on Windows it will be converted to a `wchar_t const *`. Conversely, if a `wchar_t const *` is passed to a Reverse IL Stub from native then that stub is responsible for converting it to a .NET `string`.

Reviewer: ```suggestion
- Marshal each argument into the appropriate form for the target. For example, if a .NET `string` is being passed to a native function, then the stub will convert it to a `wchar_t const *` by default on Windows. Conversely, if a `wchar_t const *` is passed to a Reverse IL Stub from native then that stub is responsible for converting it to a .NET `string`.
```

> - Marshal each argument into the appropriate form for the target. For example, if a .NET `string` is being passed to a native function then by default on Windows it will be converted to a `wchar_t const *`. Conversely, if a `wchar_t const *` is passed to a Reverse IL Stub from native then that stub is responsible for converting it to a .NET `string`.

Reviewer: Also call out attributes (`MarshalAsAttribute`) in relation to the 'appropriate form', since the list is supposed to include things that influence these actions?

- Marshal out all non-[`in`](

> https://docs.microsoft.com/dotnet/csharp/language-reference/keywords/in-parameter-modifier) by-ref arguments and the return type to the caller.

Reviewer: Should this also mention `OutAttribute`?

> - Ensure the unmanaged/managed boundary are seamless as it relates to memory leaks or corruption. For example, if passing a [`SafeHandle`][api_safehandle] derived type to a native function a leak of that handle shouldn't be possible.

Reviewer: ```suggestion
- Ensure the unmanaged/managed boundary is seamless as it relates to memory leaks or corruption. For example, if passing a [`SafeHandle`][api_safehandle] derived type to a native function, a leak of that handle shouldn't be possible.
```

An IL Stub or Reverse IL Stub will be generated in several circumstances. The following C# examples represent some of these circumstances.

**IL Stub**

The generated stub will marshal the .NET `string` `"ABCDEFG"` to the native function `to_lower()`.

```csharp
[DllImport("NativeLib.dll", EntryPoint = "to_lower")]
static extern string ToLower(string str);

> var lowered = ToLower("ABCDEFG");

Reviewer: ```suggestion
string lowered = ToLower("ABCDEFG");
```

```

Another way to call the native `to_lower()` function is using a `Delegate` type &ndash; which will also require an IL Stub.

```csharp
[UnmanagedFunctionPointer(CallingConvention.Winapi)]
delegate string ToLowerDelegate(string str);

IntPtr mod = NativeLibrary.Load("NativeLib.dll");
IntPtr fptr = NativeLibrary.GetExport(mod, "to_lower");

var toLower = Marshal.GetDelegateForFunctionPointer<ToLowerDelegate>(fptr);

> var lowered = toLower("ABCDEFG");

Reviewer: We should use the actual type instead of `var` throughout.
https://docs.microsoft.com/dotnet/csharp/programming-guide/inside-a-program/coding-conventions#implicitly-typed-local-variables

```

> The generated stub in the above examples would translate the supplied .NET `string` and convert it to the appropriate native representation. Take a moment and consider what that would imply. The .NET platform internally stores `string` values as UTF-16 encoded values. This means that if the native environment were expecting a UTF-8 or ANSI code page encoding that would need to be handled by some logic. Additionally the GC is free to move the .NET `string` if it helps optimize memory usage. The IL Stub is responsible for this logic. The `to_lower()` function returns a string though, which means the reverse operation including encoding considerations must taken into account.

Reviewer: ```suggestion
The generated stub in the above examples would convert the supplied .NET `string` to the appropriate native representation. This conversion carries some implications that are worth considering. The .NET platform internally stores `string` values as UTF-16 encoded values. This means that if the native environment were expecting a UTF-8 or ANSI code page encoding, that would need to be handled by some logic. Additionally the GC is free to move the .NET `string` if it helps optimize memory usage. The IL Stub is responsible for this logic. The `to_lower()` function also returns a string, which means the reverse operation (converting the native representation to a .NET `string`, including encoding considerations), must be taken into account.
```

> The generated stub in the above examples would translate the supplied .NET `string` and convert it to the appropriate native representation. Take a moment and consider what that would imply. The .NET platform internally stores `string` values as UTF-16 encoded values. This means that if the native environment were expecting a UTF-8 or ANSI code page encoding that would need to be handled by some logic. Additionally the GC is free to move the .NET `string` if it helps optimize memory usage. The IL Stub is responsible for this logic. The `to_lower()` function returns a string though, which means the reverse operation including encoding considerations must taken into account.

Reviewer: > Additionally the GC is free to move the .NET `string` if it helps optimize memory usage.

This sentence reads like an interloper, since the previous sentence has `handled by some logic`, this one does not say anything about logic/handling, and then the next one refers to `this logic`.

**Reverse IL Stub**

A Reverse IL Stub will be created so the native caller can provide a native string and have it marshalled to a .NET `string`.

```csharp
delegate string ToLowerDelegate(string str);

static string ToLowerManaged(string str)
{
    return str.ToLower();
}

[DllImport("NativeLib.dll", EntryPoint = "call_fptr")]
static extern void CallFptr(IntPtr fptr);

var fptr = Marshal.GetFunctionPointerForDelegate<ToLowerDelegate>(ToLowerManaged);
CallFptr(fptr);

// Native code - for the Windows platform.
//  typedef const wchar_t* (*to_lower_t)(const wchar_t* str);
//  void call_fptr(void* fptr)
//  {
//      const wchar_t* lowered = ((to_lower_t)fptr)(L"ABCDEFG");
//  }
```

> The above examples represent symmetrical operations for a function that takes a sequence of characters (that is, a string) and returns an new string that has been converted into lowercase letters. The logic of that operation is defined within the function itself but in order to correctly pass the arguments and return the value to the other environment a stub is used.

Reviewer: ```suggestion
The above examples represent symmetrical operations for a function that takes a sequence of characters (that is, a string) and returns a new string that has been converted into lowercase letters. The logic of that operation is defined within the function itself but in order to correctly pass the arguments and return the value to the other environment, a stub is used.
```

> The above examples represent symmetrical operations for a function that takes a sequence of characters (that is, a string) and returns an new string that has been converted into lowercase letters. The logic of that operation is defined within the function itself but in order to correctly pass the arguments and return the value to the other environment a stub is used.

Reviewer: Maybe move to above the **IL Stub** and **Reverse IL Stub** headings? For me, its placement makes it look like it is just for the **Reverse IL Stub** section.

Once the specific native dependency (that is, `mscoree.dll`/`ijwhost.dll`) is loaded, all managed exports exposed to native code are "thunked". This thunk provides a level of indirection between the actual managed function and the calling native code in order to load the .NET runtime on demand. The first time a managed function is called by native code, the thunk is executed and a .NET runtime loaded or the existing one confirmed to be compatible and adopted. All native exports are then populated with the appropriate managed function. Only one managed call must pay this initialization price since all thunks for an assembly will be updated when it is loaded into a .NET runtime. After a .NET runtime has loaded the assembly and the managed function thunks have been updated, the interop experience follows all the rules and principles described in the [Concepts](#concepts) section.

> Aside from the `mscoree.dll`/`ijwhost.dll` differences there are other functional discrepancies between .NET Framework and .NET Core 3.1/.NET 5+. These are captured in official documentation for transitioning C++/CLI from [.NET Framework to a newer .NET version](https://docs.microsoft.com/dotnet/core/porting/cpp-cli).

Reviewer: ```suggestion
Aside from the `mscoree.dll`/`ijwhost.dll` differences, there are other functional discrepancies between .NET Framework and .NET Core 3.1/.NET 5+. These are captured in official documentation for transitioning C++/CLI from [.NET Framework to a newer .NET version](https://docs.microsoft.com/dotnet/core/porting/cpp-cli).
```
